File: runtime/src/test/java/com/google/protobuf/TestAllTypes.java
// Copyright 2013 Square, Inc.
package com.google.protobuf;

import com.squareup.omar.Omar;
import com.squareup.protos.alltypes.AllTypesContainer;
import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import junit.framework.TestCase;

import static com.squareup.protos.alltypes.AllTypesContainer.AllTypes;

public class TestAllTypes extends TestCase {

  // Return a two-element array with a given repeated value
  private <T> List<T> array(T x) {
    return Arrays.asList(x, x);
  }

  private final AllTypes allTypes = createAllTypes();
  private final Omar omar = new Omar(AllTypesContainer.class);

  private AllTypes createAllTypes() {
    byte[] bytes = { (byte) 125, (byte) 225 };
    AllTypesContainer.AllTypes.NestedMessage nestedMessage =
        new AllTypes.NestedMessage.Builder().a(999).build();
    return new AllTypes.Builder()
        .opt_int32(111)
        .opt_uint32(112)
        .opt_sint32(113)
        .opt_fixed32(114)
        .opt_sfixed32(115)
        .opt_int64(116L)
        .opt_uint64(117L)
        .opt_sint64(118L)
        .opt_fixed64(119L)
        .opt_sfixed64(120L)
        .opt_bool(true)
        .opt_float(122.0F)
        .opt_double(123.0)
        .opt_string("124")
        .opt_bytes(bytes)
        .opt_nested_enum(AllTypes.NestedEnum.A)
        .opt_nested_message(nestedMessage)
        .req_int32(111)
        .req_uint32(112)
        .req_sint32(113)
        .req_fixed32(114)
        .req_sfixed32(115)
        .req_int64(116L)
        .req_uint64(117L)
        .req_sint64(118L)
        .req_fixed64(119L)
        .req_sfixed64(120L)
        .req_bool(true)
        .req_float(122.0F)
        .req_double(123.0)
        .req_string("124")
        .req_bytes(bytes)
        .req_nested_enum(AllTypes.NestedEnum.A)
        .req_nested_message(nestedMessage)
        .rep_int32(array(111))
        .rep_uint32(array(112))
        .rep_sint32(array(113))
        .rep_fixed32(array(114))
        .rep_sfixed32(array(115))
        .rep_int64(array(116L))
        .rep_uint64(array(117L))
        .rep_sint64(array(118L))
        .rep_fixed64(array(119L))
        .rep_sfixed64(array(120L))
        .rep_bool(array(true))
        .rep_float(array(122.0F))
        .rep_double(array(123.0))
        .rep_string(array("124"))
        .rep_bytes(array(bytes))
        .rep_nested_enum(array(AllTypes.NestedEnum.A))
        .rep_nested_message(array(nestedMessage))
        .pack_int32(array(111))
        .pack_uint32(array(112))
        .pack_sint32(array(113))
        .pack_fixed32(array(114))
        .pack_sfixed32(array(115))
        .pack_int64(array(116L))
        .pack_uint64(array(117L))
        .pack_sint64(array(118L))
        .pack_fixed64(array(119L))
        .pack_sfixed64(array(120L))
        .pack_bool(array(true))
        .pack_float(array(122.0F))
        .pack_double(array(123.0))
        .pack_string(array("124"))
        .pack_bytes(array(bytes))
        .pack_nested_enum(array(AllTypes.NestedEnum.A))
        .pack_nested_message(array(nestedMessage))
        .setExtension(AllTypesContainer.ext_pack_bool, array(true))
        .build();
  }

  public void testWrite() {
    int len = omar.getSerializedSize(allTypes);
    assertEquals(AllTypesData.expectedOutput.length, len);
    byte[] output = new byte[len];
    omar.writeTo(allTypes, output, 0, len);
    for (int i = 0; i < output.length; i++) {
      assertEquals("Byte " + i, AllTypesData.expectedOutput[i], output[i] & 0xff);
    }
  }

  public void testRead() throws IOException {
    byte[] data = new byte[AllTypesData.expectedOutput.length];
    omar.writeTo(allTypes, data, 0, data.length);
    AllTypes parsed = omar.parseFrom(AllTypes.class, data);
    assertEquals(allTypes, parsed);
  }

  public void testReadNoExtension() throws IOException {
    byte[] data = new byte[AllTypesData.expectedOutput.length];
    omar.writeTo(allTypes, data, 0, data.length);
    AllTypes parsed = new Omar().parseFrom(AllTypes.class, data);
    assertFalse(allTypes.equals(parsed));
  }

  public void testReadNonPacked() throws IOException {
    byte[] data = new byte[AllTypesData.nonPacked.length];
    for (int i = 0; i < AllTypesData.nonPacked.length; i++) {
      data[i] = (byte) AllTypesData.nonPacked[i];
    }
    AllTypes parsed = omar.parseFrom(AllTypes.class, data);
    assertEquals(allTypes, parsed);
  }

  public void testToString() throws IOException {
    byte[] data = new byte[AllTypesData.expectedOutput.length];
    omar.writeTo(allTypes, data, 0, data.length);
    AllTypes parsed = omar.parseFrom(AllTypes.class, data);
    assertEquals(AllTypesData.expectedToString, parsed.toString());
  }
}

File: wire-tests/src/commonTest/kotlin/com/squareup/wire/TestAllTypes.kt
/*
 * Copyright (C) 2013 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire

import assertk.assertions.message
import com.squareup.wire.TestAllTypesData.list
import com.squareup.wire.TestAllTypesData.message
import com.squareup.wire.protos.kotlin.alltypes.AllTypes
import kotlin.test.Test
import kotlin.test.assertEquals
import kotlin.test.assertNull
import okio.Buffer
import okio.ByteString

class TestAllTypes {
  private val adapter = AllTypes.ADAPTER

  private val allTypes = TestAllTypesData.allTypes

  @Test fun testHashCodes() {
    val message = message()
    val messageHashCode = message.hashCode()
    assertEquals(allTypes.hashCode(), messageHashCode)
  }

  @Test fun testBuilder() {
    var builder = message()
    val nestedMessage = AllTypes.NestedMessage(a = 999)

    assertEquals(111, builder.opt_int32)
    assertEquals(112, builder.opt_uint32)
    assertEquals(113, builder.opt_sint32)
    assertEquals(114, builder.opt_fixed32)
    assertEquals(115, builder.opt_sfixed32)
    assertEquals(116L, builder.opt_int64)
    assertEquals(117L, builder.opt_uint64)
    assertEquals(118L, builder.opt_sint64)
    assertEquals(119L, builder.opt_fixed64)
    assertEquals(120L, builder.opt_sfixed64)
    assertEquals(true, builder.opt_bool)
    assertEquals(122.0f, builder.opt_float)
    assertEquals(123.0, builder.opt_double)
    assertEquals("124", builder.opt_string)
    assertEquals(ByteString.of(125.toByte(), 225.toByte()), builder.opt_bytes)
    assertEquals(AllTypes.NestedEnum.A, builder.opt_nested_enum)
    assertEquals(nestedMessage, builder.opt_nested_message)

    assertEquals(111, builder.req_int32)
    assertEquals(112, builder.req_uint32)
    assertEquals(113, builder.req_sint32)
    assertEquals(114, builder.req_fixed32)
    assertEquals(115, builder.req_sfixed32)
    assertEquals(116L, builder.req_int64)
    assertEquals(117L, builder.req_uint64)
    assertEquals(118L, builder.req_sint64)
    assertEquals(119L, builder.req_fixed64)
    assertEquals(120L, builder.req_sfixed64)
    assertEquals(true, builder.req_bool)
    assertEquals(122.0f, builder.req_float)
    assertEquals(123.0, builder.req_double)
    assertEquals("124", builder.req_string)
    assertEquals(ByteString.of(125.toByte(), 225.toByte()), builder.req_bytes)
    assertEquals(AllTypes.NestedEnum.A, builder.req_nested_enum)
    assertEquals(nestedMessage, builder.req_nested_message)

    assertEquals(2, builder.rep_int32.size)
    assertEquals(111, builder.rep_int32[0])
    assertEquals(111, builder.rep_int32[1])
    assertEquals(2, builder.rep_uint32.size)
    assertEquals(112, builder.rep_uint32[0])
    assertEquals(112, builder.rep_uint32[1])
    assertEquals(2, builder.rep_sint32.size)
    assertEquals(113, builder.rep_sint32[0])
    assertEquals(113, builder.rep_sint32[1])
    assertEquals(2, builder.rep_fixed32.size)
    assertEquals(114, builder.rep_fixed32[0])
    assertEquals(114, builder.rep_fixed32[1])
    assertEquals(2, builder.rep_sfixed32.size)
    assertEquals(115, builder.rep_sfixed32[0])
    assertEquals(115, builder.rep_sfixed32[1])
    assertEquals(2, builder.rep_int64.size)
    assertEquals(116L, builder.rep_int64[0])
    assertEquals(116L, builder.rep_int64[1])
    assertEquals(2, builder.rep_uint64.size)
    assertEquals(117L, builder.rep_uint64[0])
    assertEquals(117L, builder.rep_uint64[1])
    assertEquals(2, builder.rep_sint64.size)
    assertEquals(118L, builder.rep_sint64[0])
    assertEquals(118L, builder.rep_sint64[1])
    assertEquals(2, builder.rep_fixed64.size)
    assertEquals(119L, builder.rep_fixed64[0])
    assertEquals(119L, builder.rep_fixed64[1])
    assertEquals(2, builder.rep_sfixed64.size)
    assertEquals(120L, builder.rep_sfixed64[0])
    assertEquals(120L, builder.rep_sfixed64[1])
    assertEquals(2, builder.rep_bool.size)
    assertEquals(true, builder.rep_bool[0])
    assertEquals(true, builder.rep_bool[1])
    assertEquals(2, builder.rep_float.size)
    assertEquals(122.0f, builder.rep_float[0])
    assertEquals(122.0f, builder.rep_float[1])
    assertEquals(2, builder.rep_double.size)
    assertEquals(123.0, builder.rep_double[0])
    assertEquals(123.0, builder.rep_double[1])
    assertEquals(2, builder.rep_string.size)
    assertEquals("124", builder.rep_string[0])
    assertEquals("124", builder.rep_string[1])
    assertEquals(2, builder.rep_bytes.size)
    assertEquals(ByteString.of(125.toByte(), 225.toByte()), builder.rep_bytes[0])
    assertEquals(ByteString.of(125.toByte(), 225.toByte()), builder.rep_bytes[1])
    assertEquals(2, builder.rep_nested_enum.size)
    assertEquals(AllTypes.NestedEnum.A, builder.rep_nested_enum[0])
    assertEquals(AllTypes.NestedEnum.A, builder.rep_nested_enum[1])
    assertEquals(2, builder.rep_nested_message.size)
    assertEquals(nestedMessage, builder.rep_nested_message[0])
    assertEquals(nestedMessage, builder.rep_nested_message[1])

    assertEquals(2, builder.pack_int32.size)
    assertEquals(111, builder.pack_int32[0])
    assertEquals(111, builder.pack_int32[1])
    assertEquals(2, builder.pack_uint32.size)
    assertEquals(112, builder.pack_uint32[0])
    assertEquals(112, builder.pack_uint32[1])
    assertEquals(2, builder.pack_sint32.size)
    assertEquals(113, builder.pack_sint32[0])
    assertEquals(113, builder.pack_sint32[1])
    assertEquals(2, builder.pack_fixed32.size)
    assertEquals(114, builder.pack_fixed32[0])
    assertEquals(114, builder.pack_fixed32[1])
    assertEquals(2, builder.pack_sfixed32.size)
    assertEquals(115, builder.pack_sfixed32[0])
    assertEquals(115, builder.pack_sfixed32[1])
    assertEquals(2, builder.pack_int64.size)
    assertEquals(116L, builder.pack_int64[0])
    assertEquals(116L, builder.pack_int64[1])
    assertEquals(2, builder.pack_uint64.size)
    assertEquals(117L, builder.pack_uint64[0])
    assertEquals(117L, builder.pack_uint64[1])
    assertEquals(2, builder.pack_sint64.size)
    assertEquals(118L, builder.pack_sint64[0])
    assertEquals(118L, builder.pack_sint64[1])
    assertEquals(2, builder.pack_fixed64.size)
    assertEquals(119L, builder.pack_fixed64[0])
    assertEquals(119L, builder.pack_fixed64[1])
    assertEquals(2, builder.pack_sfixed64.size)
    assertEquals(120L, builder.pack_sfixed64[0])
    assertEquals(120L, builder.pack_sfixed64[1])
    assertEquals(2, builder.pack_bool.size)
    assertEquals(true, builder.pack_bool[0])
    assertEquals(true, builder.pack_bool[1])
    assertEquals(2, builder.pack_float.size)
    assertEquals(122.0f, builder.pack_float[0])
    assertEquals(122.0f, builder.pack_float[1])
    assertEquals(2, builder.pack_double.size)
    assertEquals(123.0, builder.pack_double[0])
    assertEquals(123.0, builder.pack_double[1])
    assertEquals(2, builder.pack_nested_enum.size)
    assertEquals(AllTypes.NestedEnum.A, builder.pack_nested_enum[0])
    assertEquals(AllTypes.NestedEnum.A, builder.pack_nested_enum[1])

    assertEquals(true, builder.ext_opt_bool)
    assertEquals(list(true), builder.ext_rep_bool)
    assertEquals(list(true), builder.ext_pack_bool)

    builder = builder.copy(
      ext_opt_bool = false,
      ext_rep_bool = list(false),
      ext_pack_bool = list(false),
    )

    assertEquals(false, builder.ext_opt_bool)
    assertEquals(list(false), builder.ext_rep_bool)
    assertEquals(list(false), builder.ext_pack_bool)
  }

  @Test fun testWrite() {
    val output = adapter.encode(allTypes)
    assertEquals(TestAllTypesData.expectedOutput, ByteString.of(*output))
    assertEquals(TestAllTypesData.expectedOutput.size, output.size)
  }

  @Test fun testWriteSource() {
    val sink = Buffer()
    adapter.encode(sink, allTypes)
    assertEquals(TestAllTypesData.expectedOutput, sink.readByteString())
  }

  @Test fun testWriteBytes() {
    val output = adapter.encode(allTypes)
    assertEquals(TestAllTypesData.expectedOutput.size, output.size)
    assertEquals(TestAllTypesData.expectedOutput, ByteString.of(*output))
  }

  @Test fun testReadSource() {
    val data = adapter.encode(allTypes)
    val input = Buffer().write(data)
    val parsed = adapter.decode(input)
    assertEquals(allTypes, parsed)
    assertEquals(true, allTypes.ext_opt_bool)
    assertEquals(list(true), allTypes.ext_rep_bool)
    assertEquals(list(true), allTypes.ext_pack_bool)
  }

  @Test fun testReadBytes() {
    val data = adapter.encode(allTypes)
    val parsed = adapter.decode(data)
    assertEquals(allTypes, parsed)
    assertEquals(true, allTypes.ext_opt_bool)
    assertEquals(list(true), allTypes.ext_rep_bool)
    assertEquals(list(true), allTypes.ext_pack_bool)
  }

  @Test fun testReadLongMessages() {
    val allTypes = message(50)
    val data = adapter.encode(allTypes)
    val parsed = adapter.decode(data)
    assertEquals(allTypes, parsed)
    assertEquals(true, allTypes.ext_opt_bool)
    assertEquals(list(true, 50), allTypes.ext_rep_bool)
    assertEquals(list(true, 50), allTypes.ext_pack_bool)
  }

  @Test fun testReadNoExtension() {
    val data = adapter.encode(allTypes)
    val parsed = AllTypes.ADAPTER.decode(data)
    assertEquals(allTypes, parsed)
  }

  @Test fun testReadNonPacked() {
    val parsed = adapter.decode(Buffer().write(TestAllTypesData.nonPacked))
    assertEquals(allTypes, parsed)
  }

  @IgnoreJs // https://youtrack.jetbrains.com/issue/KT-35078
  @Test
  fun testToString() {
    val data = adapter.encode(allTypes)
    val parsed = adapter.decode(data)
    assertEquals(TestAllTypesData.expectedToString, parsed.toString())
  }

  @Test fun testDefaults() {
    assertEquals(true, AllTypes.DEFAULT_DEFAULT_BOOL)
    // original: "<c-cedilla>ok\a\b\f\n\r\t\v\1\01\001\17\017\176\x1\x01\x11\X1\X01\X11g<u umlaut>zel"
    assertEquals(
      "çok\u0007\b\u000C\n\r\t\u000b\u0001\u0001\u0001\u000f\u000f~\u0001\u0001\u0011" +
        "\u0001\u0001\u0011güzel",
      AllTypes.DEFAULT_DEFAULT_STRING,
    )
  }

  @Test fun testEnums() {
    assertEquals(AllTypes.NestedEnum.A, AllTypes.NestedEnum.fromValue(1))
    assertNull(AllTypes.NestedEnum.fromValue(10))
    assertEquals(1, AllTypes.NestedEnum.A.value)
  }

  @Test fun testSkipGroup() {
    val data = ByteArray(TestAllTypesData.expectedOutput.size + 27)
    arraycopy(TestAllTypesData.expectedOutput.toByteArray(), 0, data, 0, 17)
    var index = 17
    data[index++] = 0xa3.toByte() // start group, tag = 20, type = 3
    data[index++] = 0x01.toByte()
    data[index++] = 0x08.toByte() // tag = 1, type = 0 (varint)
    data[index++] = 0x81.toByte()
    data[index++] = 0x82.toByte()
    data[index++] = 0x6f.toByte()
    data[index++] = 0x21.toByte() // tag = 2, type = 1 (fixed64)
    data[index++] = 0x01.toByte()
    data[index++] = 0x02.toByte()
    data[index++] = 0x03.toByte()
    data[index++] = 0x04.toByte()
    data[index++] = 0x05.toByte()
    data[index++] = 0x06.toByte()
    data[index++] = 0x07.toByte()
    data[index++] = 0x08.toByte()
    data[index++] = 0x1a.toByte() // tag = 3, type = 2 (length-delimited)
    data[index++] = 0x03.toByte() // length = 3
    data[index++] = 0x01.toByte()
    data[index++] = 0x02.toByte()
    data[index++] = 0x03.toByte()
    data[index++] = 0x25.toByte() // tag = 4, type = 5 (fixed32)
    data[index++] = 0x01.toByte()
    data[index++] = 0x02.toByte()
    data[index++] = 0x03.toByte()
    data[index++] = 0x04.toByte()
    data[index++] = 0xa4.toByte() // end group, tag = 20, type = 4
    data[index++] = 0x01.toByte()
    arraycopy(
      TestAllTypesData.expectedOutput.toByteArray(),
      17,
      data,
      index,
      TestAllTypesData.expectedOutput.size - 17,
    )
    val parsed = adapter.decode(data)
    assertEquals(allTypes, parsed)
  }

  private fun arraycopy(src: ByteArray, srcPos: Int, dest: ByteArray, destPos: Int, length: Int) {
    for (offset in 0 until length) {
      dest[destPos + offset] = src[srcPos + offset]
    }
  }
}
